File: shared-upnp/src/main/kotlin/com/m3sv/plainupnp/upnp/android/DlnaSaxParser.java
package com.m3sv.plainupnp.upnp.android;

import org.seamless.xml.SAXParser;
import org.xml.sax.XMLReader;

import javax.xml.parsers.SAXParserFactory;

public class DlnaSaxParser extends SAXParser {
    @Override
    protected XMLReader create() {
        try {
            SAXParserFactory factory = SAXParserFactory.newInstance();

            // Configure factory to prevent XXE attacks
            factory.setFeature("http://xml.org/sax/features/external-general-entities", false);
            factory.setFeature("http://xml.org/sax/features/external-parameter-entities", false);

            XMLReader xmlReader = factory.newSAXParser().getXMLReader();
            xmlReader.setErrorHandler(getErrorHandler());
            return xmlReader;
        } catch (Exception ex) {
            throw new RuntimeException(ex);
        }
    }
}

File: upnp/src/main/kotlin/com/m3sv/plainupnp/upnp/android/DlnaSaxParser.kt
package com.m3sv.plainupnp.upnp.android

import org.seamless.xml.SAXParser
import org.xml.sax.XMLReader
import javax.xml.parsers.SAXParserFactory

class DlnaSaxParser : SAXParser() {
    override fun create(): XMLReader = try {
        SAXParserFactory
            .newInstance()
            .apply {
                // Configure factory to prevent XXE attacks
                setFeature("http://xml.org/sax/features/external-general-entities", false)
                setFeature("http://xml.org/sax/features/external-parameter-entities", false)
            }
            .newSAXParser()
            .xmlReader
    } catch (ex: Exception) {
        throw RuntimeException(ex)
    }
}
